### ISSUE TITLE
SCRAM-SHA-1(-PLUS) + SCRAM-SHA-256(-PLUS) supports

### ISSUE THREAD

**Issue by Neustradamus** [roundcube/roundcubemail#6917]
"When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant, and SHA-256 variants [RFC7677] SHOULD be preferred over SHA-1 variants [RFC5802]".

Can you add support for?
- SCRAM-SHA-1(-PLUS):
-- https://tools.ietf.org/html/rfc5802
-- https://tools.ietf.org/html/rfc6120

- SCRAM-SHA-256(-PLUS):
-- https://tools.ietf.org/html/rfc7677 since 2015-11-02
-- https://tools.ietf.org/html/rfc8600 since 2019-06-21: https://mailarchive.ietf.org/arch/msg/ietf-announce/suJMmeMhuAOmGn_PJYgX5Vm8lNA

I add SCRAM-SHA-512(-PLUS): https://xmpp.org/extensions/inbox/hash-recommendations.html

Linked to:
- https://github.com/scram-xmpp/info/issues/1

**Comment by alecpl** [comment]
Auth_SASL has some SCRAM support so it should be doable, but I will not give it a high prio. Any help appreciated.

**Comment by Neustradamus** [comment]
- There is: Auth_SASL2 : https://pear.php.net/package/Auth_SASL2
- Maybe good to compare with Auth_SASL: https://pear.php.net/package/Auth_SASL

It can be ok for 1.4?

**Comment by alecpl** [comment]
There's only one difference between these two libs. They handle errors differently. SO, there's no reason to switch.

**Comment by Neustradamus** [comment]
What is missing for SCRAM in 1.4 RC1 for 1.4 release?

Note: SCRAM is for LDAP too.

**Comment by alecpl** [comment]
1. Make sure pear/Auth_SASL has everything you need. It might not support *-PLUS variants, but I didn't look at this.
2. Add SCRAM auth support to rcube_imap_generic.
3. Add SCRAM auth support to pear/Net_Sieve.
4. Add SCRAM auth support to pear/Net_SMTP.
5. Add SCRAM auth support to kolab/Net_LDAP3.

**Comment by Neustradamus** [comment]
Yes, I think it is not supported yet for -PLUS variants, but it is already good point :)

**Comment by Neustradamus** [comment]
@alecpl: Any news?
I have seen that Roundcube 1.4 is out!

**Comment by alecpl** [comment]
I'm not working and not plan to work on this in foreseable future. It's not a prio for me. Pull requests appreciated.

**Comment by Neustradamus** [comment]
@alecpl, @kanarip: Can you look for kolab/Net_LDAP3 and SCRAM?
I have looked here: https://git.kolab.org/diffusion/PNL/history/master/

@alecpl: Your are here too :)
- https://github.com/pear/Net_Sieve/commits/master
Why you do not accept ticket here?

I have done a ticket for pear/Net_SMTP: https://github.com/pear/Net_SMTP/issues/57

For pear/Auth_SASL, all except -PLUS variants are good.

**Comment by Neustradamus** [comment]
@alecpl: Can you add "Issues" section (or it must be owner of PEAR) in https://github.com/pear/Net_Sieve?

Have you progressed on SCRAM?
To have a best security?

Note, after SCRAM-SHA-1(-PLUS):
- GNU SASL (gsasl) supports SCRAM-SHA-256(-PLUS) since 1.9.1: http://git.savannah.gnu.org/gitweb/?p=gsasl.git;a=blob;f=NEWS;hb=HEAD
- Dovecot supports SCRAM-SHA-256(-PLUS) since 2.3.10: https://doc.dovecot.org/configuration_manual/authentication/authentication_mechanisms/
- Cyrus SASL/IMAP supports SCRAM-SHA-256(-PLUS) and more since 2.1.27: https://www.cyrusimap.org/sasl/sasl/authentication_mechanisms.html
- ...

**Comment by Neustradamus** [comment]
@alecpl: SquirrelMail has SCRAM, it is possible to add in Roundcube?
- https://github.com/RealityRipple/squirrelmail

**Comment by Neustradamus** [comment]
Any news for 1.5 ?

**Comment by Neustradamus** [comment]
@alecpl: Have you progressed on it?

**Comment by Neustradamus** [comment]
Dear @roundcube team, @alecpl, @thomascube, @vbenincasa,

Happy New Year 2022!

Have you looked for SCRAM-SHA-* supports?
The first step is without -PLUS variants.

Thanks in advance.

cc: @jparise

**Comment by Neustradamus** [comment]
I have forgotten to inform you, SCRAM-SHA-256/SCRAM-SHA-256-PLUS are in the new IMAP, I have added the last RFC in the description: RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2:
- https://tools.ietf.org/html/rfc9051

I wish you a good reading ^^

**Comment by pkolmann** [comment]
Please consider when implementing SCRAM-SHA-* that the whole SCRAM-algorithm should run on the Client, not the PHP on the server.

UseCase: Company Firewall with deep packet inspection should not get the plaintext password in the HTTPS traffic...

**Comment by Neustradamus** [comment]
@roundcube team, @alecpl: Net_SMTP (with Auth_SASL/Auth_SASL2) has now SCRAM-SHA-* features, old mechanisms have been deprecated.
Not yet -PLUS variants but a big progress done by @schengawegga and you can use.

Can you look for rcube_imap_generic, Net_Sieve, Net_LDAP3?

**Comment by Neustradamus** [comment]
Dear @karlitschek, @owncloud team, @roundcube team,

Security is the number 1?

Can you confirm it and add SCRAM-SHA-* and SCRAM-SHA-*-PLUS supports in Roundcube, Nextcloud and deps?

Thanks in advance.

Roundcube, linked to (comment from @alecpl):
- https://github.com/roundcube/roundcubemail/issues/6917#issuecomment-529422612

Nextcloud, linked to:
- https://github.com/search?q=org%3Anextcloud+scram&type=issues
- https://github.com/nextcloud/mail/issues/3146
- https://github.com/nextcloud/server/issues/39732
- https://github.com/symfony/symfony/issues/51292
- https://github.com/nextcloud/mail/pull/4377
- https://github.com/nextcloud/mail/blob/098d808c7c1a9af47a83a9fb27446a092092de1c/CHANGELOG.md?plain=1#L763
- https://bugs.horde.org/ticket/15146

Linked to:
- https://github.com/scram-sasl/info/issues/1

**Comment by Neustradamus** [comment]
@alecpl: Can you look for rcube_imap_generic, Net_Sieve, Net_LDAP3?

It is already in pear/Auth_SASL, pear/Net_SMTP:
- https://github.com/pear/Auth_SASL/releases
- https://github.com/pear/Auth_SASL2/releases
- https://github.com/pear/Net_SMTP/releases

Thanks in advance.

cc: @schengawegga.

**Comment by Neustradamus** [comment]
Dear @karlitschek, @nextcloud team.

I wish you a Happy New Year 2024!

Can you add SCRAM supports in Roundcube and in Nextcloud?

**Comment by pabzm** [comment]
As @alecpl said [five years ago](https://github.com/roundcube/roundcubemail/issues/6917#issuecomment-554613463): Pull requests welcome. Roundcube is very much a community project. If you care for this feature:

* find more people that want it. You're the only one since years who asks for it. Repeatedly tagging single people to push for it won't make it happen.
* find someone who implements it. The dev team has limited time, too, we can't do all the work.

**Comment by Neustradamus** [comment]
@schengawegga has done a lot of work in PEAR repositories but I do not know about Roundcube.
Hope an answer.

**Comment by Neustradamus** [comment]
@Danelif, @josaphatim: You have done a good part in Cypht, it is possible to look Roundcube?
